lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js

```diff
@@ -0,0 +1,149 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var isComplexTypedArray = require( '@stdlib/array/base/assert/is-complex-typed-array' );
+var isAccessorArray = require( '@stdlib/array/base/assert/is-accessor-array' );
+var resolveGetter = require( '@stdlib/array/base/resolve-getter' );
+var isSameValue = require( '@stdlib/assert/is-same-value' );
+
+
+// FUNCTIONS //
+
+/**
+* Count the number of elements that are equal to a given value in an indexed array.
+*
+* @private
+* @param {Collection} x - input array
+* @param {*} value - given value
+* @returns {NonNegativeInteger} number of elements that are equal to the given value
+*
+* @example
+* var x = [ 0, 1, 0, 1 ];
+*
+* var n = indexed( x, 1 );
+* // returns 2
+*/
+function indexed( x, value ) {
+	var n;
+	var i;
+
+	n = 0;
+	for ( i = 0; i < x.length; i++ ) {
+		if ( isSameValue(x[i], value) ) {
+			n += 1;
+		}
+	}
+	return n;
+}
+
+/**
+* Count the number of elements that are equal to a given value in an accessor array.
+*
+* @private
+* @param {Collection} x - input array
+* @param {*} value - given value
+* @returns {NonNegativeInteger} number of elements that are equal to the given value
+*
+* @example
+* var toAccessorArray = require( '@stdlib/array/base/to-accessor-array' );
+*
+* var x = toAccessorArray( [ 0, 1, 0, 1 ] );
+*
+* var n = accessors( x, 1 );
+* // returns 2
+*/
+function accessors( x, value ) {
+	var get;
+	var n;
+	var i;
+
+	get = resolveGetter( x );
+
+	n = 0;
+	for ( i = 0; i < x.length; i++ ) {
+		if ( isSameValue(get( x, i ), value) ) {
+			n += 1;
+		}
+	}
+	return n;
+}
+
+/**
+* Count the number of elements that are equal to a given value in a complex array.
+*
+* @private
+* @param {Collection} x - input array
+* @param {*} value - given value
+* @returns {NonNegativeInteger} number of elements that are equal to the given value
+*
+* @example
+* var Complex128 = require( '@stdlib/complex/float64' );
+* var Complex128Array = require( '@stdlib/array/complex128' );
+*
+* var x = new Complex128Array( [ 1.0, 2.0, 0.0, 0.0, 3.0, 4.0, 0.0, 0.0 ] );
+*
+* var n = complex( x, new Complex128( 1.0, 2.0 ) );
+* // returns 1
+*/
+function complex( x, value ) {
+	var n;
+	var i;
+
+	n = 0;
+	for ( i = 0; i < x.length; i++ ) {
+		if ( isSameValue(x.get(i), value) ) {
```

Comment:
Hmm...this is a trickier one. (1) We probably want to avoid complex number object comparison, as it is slower. `x.get( i )`, will return a new complex number instance, and that construction will be relatively expensive. In which case, we'd like to avoid if possible. (2) We'd want to consider whether a real-valued `value` (e.g., `9.0`) should compare as equal to `9 + 0j` (i.e., a real-number having an implied zero for its imaginary component). Given that [`@stdlib/assert/is-same-value`](https://github.com/stdlib-js/stdlib/blob/64931c5fdb4abc3a339b619658d3b5acc089e556/lib/node_modules/%40stdlib/assert/is-same-value/lib/main.js) requires that both be complex number objects, I think the answer is no.

Nevertheless, we can improve this implementation in the following ways:

1. Before L111, we should check whether `value` is complex-like. If not, we should return `0`, as we know that only complex numbers can compare as equal.
2. If `value` is complex-like, we should decompose into real and imaginary components (e.g., using `@stdlib/complex/real`/`@stdlib/complex/realf`).
3. We should reinterpret the complex number array as a real-valued array using `@stdlib/strided/base/reinterpret-complex`.
4. The loop condition should be `i = 0; i < x.length; i += 2`
5. L113 should then be something along the lines of `if ( isSameValue( x[ i ], real ) && isSameValue( x[ i+1 ], imag ) {`.

This should avoid complex number instance creation and be much faster.

That make sense, @skoriop?